<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### Table of Contents

-   [formatAttributes][1]
-   [withEither][2]
-   [SelectionOptions][3]
-   [hasBlastDb][4]
-   [mapFunction][5]
-   [DB_TYPES][6]
-   [removeBlastDb][7]
-   [revcomp][8]
-   [updateTrackPermissions][9]
-   [updateSampleInfo][10]
-   [submitBlastJob][11]
-   [makeBlastDb][12]
-   [tableColumnDataTracker][13]
-   [formatFasta][14]
-   [isLoading][15]
-   [round][16]
-   [determineSeqType][17]
-   [unescape][18]
-   [reverseComplement][19]
-   [dataTracker][20]
-   [reduceFunction][21]
-   [Loading][22]
-   [Interval][23]
-   [log][24]
-   [withConditionalRendering][25]
-   [withConditionalRendering][26]
-   [translate][27]
-   [translate][28]
-   [SequenceInput][29]
-   [GeneTableOptions][30]
-   [hasNoResults][31]
-   [NoResults][32]
-   [lockGene][33]
-   [GeneModel][34]
-   [unlockGene][35]
-   [blastDataTracker][36]
-   [getMultipleGeneSequences][37]
-   [BlastResult][38]
-   [GeneTableRow][39]
-   [getGeneSequences][40]
-   [BLASTTYPES][41]
-   [getReferenceSequences][42]

## formatAttributes

Format gff3 attribute column into key:value object.

**Parameters**

-   `attributeString` **[String][43]** Raw gff3 attribute column string

Returns **[Object][44]** Key:value pairs of attributes.

## withEither

Higher order component for conditional rendering, to be used with compose
[https://www.robinwieruch.de/gentle-introduction-higher-order-components/][45]

**Parameters**

-   `conditionalRenderingFn` **[Function][46]** Function returning a boolean for condional rendering
-   `EitherComponent` **[Function][46]** Component to render if conditionalRenderingFn evaluates to true

Returns **[Function][46]** 

## SelectionOptions

[description]

**Parameters**

-   `$0` **[Object][44]** 
    -   `$0.selectedGenes`  
    -   `$0.selectedAllGenes`  
    -   `$0.toggleDownloadDialog`  

Returns **[Function][46]** [description]

## hasBlastDb

hasBlastDb validated method: checks whether blast databases exist for a given annotation track

Returns **Bool** jobqueue

## mapFunction

Map function for mongodb mapreduce

## DB_TYPES

Keep track of what blast commands should use which databases

Type: [Object][44]

## removeBlastDb

hasBlastDb validated method: checks whether blast databases exist for a given annotation track

Returns **Bool** jobqueue

## revcomp

Reverse complement a DNA string

**Parameters**

-   `seq` **[String][43]** [String representing DNA constisting of alphabet AaCcGgTtNn]

Returns **[String][43]** [String representing DNA constisting of alphabet AaCcGgTtNn, reverse complement of input]

## updateTrackPermissions

submitBlastJob validated method: submits makeblastdb job to jobqueue, call this from the client

Returns **[String][43]** jobId of the makeblastdb job

## updateSampleInfo

updateSampleInfo validated method: Update transcriptome information and groups

Returns **[String][43]** jobId of the makeblastdb job

## submitBlastJob

submitBlastJob validated method: submits makeblastdb job to jobqueue, call this from the client

Returns **[String][43]** jobId of the makeblastdb job

## makeBlastDb

makeBlastDb validated method: submits makeblastdb job to jobqueue, call this from the client

Returns **[String][43]** jobId of the makeblastdb job

## tableColumnDataTracker

[description]

**Parameters**

-   `props` **\[type]** [description]

Returns **\[type]** [description]

## formatFasta

formatFasta description

**Parameters**

-   `query` **[Object][44]** Database query to select genes
-   `sequenceType` **[String][43]** One of 'protein' or 'nucleotide'

Returns **[Array][47]** Array of fasta formatted coding sequences

## isLoading

Helper function to see if reactive dataTracker is still loading

**Parameters**

-   `$0` **[Object][44]** 
    -   `$0.loading`  

Returns **[Boolean][48]** 

## round

[https://stackoverflow.com/a/46854785/6573438][49]

**Parameters**

-   `x` **[Number][50]** Number to be rounded
-   `n` **[Number][50]** Floating point precision

Returns **[Number][50]** Rounded Number

## determineSeqType

Function to determine whether a given sequence string is DNA or protein

**Parameters**

-   `seq` **[String][43]** Input sequence, unknown if DNA or protein

Returns **[String][43]** Either 'Nucleotide' or 'Protein'

## unescape

Override the default querystring unescape function to be able to parse commas correctly in gff attributes

**Parameters**

-   `uri`  
-   `String`  

Returns **[String][43]** 

## reverseComplement

Reverse complement a DNA string

**Parameters**

-   `seq` **[String][43]** String representing DNA constisting of alphabet AaCcGgTtNn

Returns **[String][43]** String representing DNA constisting of alphabet AaCcGgTtNn, reverse complement of input

## dataTracker

Reactive Meteor tracker for GeneTable component

**Parameters**

-   `$0` **[Object][44]** 
    -   `$0.query`  
    -   `$0.scrollLimit`  
    -   `$0.selectedGenes`  
    -   `$0.updateSelection`  
    -   `$0.selectedAll`  

Returns **[Object][44]** [description]

## reduceFunction

Reduce function for mongodb mapreduce

**Parameters**

-   `_key` **[String][43]** [description]
-   `values` **[Array][47]** [description]

Returns **[Object][44]** [description]

## Loading

Simple React Loading component

Returns **[Function][46]** React Loading component

## Interval

Interval Class containing a single genomic interval. Every line in a gff3 file is an interval

Type: [Interval][51]

## log

If the query has not been used before, create a file from it. 
Otherwise use the cached file and increment the download count.
Return md5 hash of download query as download url

## withConditionalRendering

[withConditionalRendering description]

Type: \[type]

## withConditionalRendering

[withConditionalRendering description]

Type: \[type]

## translate

Convert a DNA string into a amino acid string

**Parameters**

-   `seq` **[String][43]** String representing DNA constisting of alphabet ACGTN

Returns **[String][43]** \[tring representing the amino acid complement of input string

## translate

Convert a DNA string into a amino acid string

**Parameters**

-   `seq` **[String][43]** String representing DNA constisting of alphabet ACGTN

Returns **[String][43]** String representing the amino acid complement of input string

## SequenceInput

Textarea input field to input sequences for blasting

**Parameters**

-   `props` **[Object][44]** [description]

Returns **[SequenceInput][52]** [description]

## GeneTableOptions

**Extends React.PureComponent**

**Parameters**

-   `props`  

## hasNoResults

[description]

**Parameters**

-   `props` **\[type]** [description]

Returns **\[type]** [description]

## NoResults

[description]

**Parameters**

-   `props` **\[type]** [description]
    -   `props.selectedColumns`  
    -   `props.props` **...any** 

Returns **\[type]** [description]

## lockGene

If a gene is being edited it is locked so other users can not edit it as well.

**Parameters**

-   `geneId` **[String][43]** Gene ID of the gene that should be locked for editing

Returns **[undefined][53]** No return value

## GeneModel

Genemodel Class containing all intervals for a single gene.

Type: [GeneModel][54]

## unlockGene

This unlocks a gene from being blocked during editing. 
A gene should only be unlocked by the person that locked it

**Parameters**

-   `geneId` **[String][43]** Gene ID of the gene that should be locked for editing

Returns **[undefined][53]** No return value

## blastDataTracker

Meteor reactive data tracker to fetch blast job results based on url

**Parameters**

-   `props` **[Object][44]** input props passed to React component

Returns **[Object][44]** Modified props based on Meteor reactive data

## getMultipleGeneSequences

Get nucleotide and protein sequences of all transcripts for a multiple genes (using CDS subfeatures)

**Parameters**

-   `genes` **[Array][47]&lt;gene>** Array of Object representing gene-models

Returns **[Array][47]** Array with objects, where each object has a transcriptId, 
                       nucleotide sequence and protein sequence field

## BlastResult

**Extends React.Component**

## GeneTableRow

[description]

**Parameters**

-   `$0` **[Object][44]** 
    -   `$0.gene`  
    -   `$0.selectedColumns`  
    -   `$0.selectedAllGenes`  
    -   `$0.selectedGenes`  
    -   `$0.updateSelection`  
    -   `$0.attributes`  
    -   `$0.props` **...any** 

Returns **\[type]** [description]

## getGeneSequences

Get nucleotide and protein sequences of all transcripts for a single gene (using CDS subfeatures)

**Parameters**

-   `gene` **[Object][44]** Object representing a gene-model

Returns **[Array][47]** Array with objects, where each object has a transcriptId, 
                       nucleotide sequence and protein sequence field

## BLASTTYPES

Hard coded map of sequence types to blast database types to select the appropriate blast program

Type: [Object][44]

## getReferenceSequences

Retrieve

**Parameters**

-   `referenceName` **[String][43]** Reference sequence name

Returns **[Object][44]** Object with all sequences belonging to a reference. Keys are header and values are DNA sequence strings

[1]: #formatattributes

[2]: #witheither

[3]: #selectionoptions

[4]: #hasblastdb

[5]: #mapfunction

[6]: #db_types

[7]: #removeblastdb

[8]: #revcomp

[9]: #updatetrackpermissions

[10]: #updatesampleinfo

[11]: #submitblastjob

[12]: #makeblastdb

[13]: #tablecolumndatatracker

[14]: #formatfasta

[15]: #isloading

[16]: #round

[17]: #determineseqtype

[18]: #unescape

[19]: #reversecomplement

[20]: #datatracker

[21]: #reducefunction

[22]: #loading

[23]: #interval

[24]: #log

[25]: #withconditionalrendering

[26]: #withconditionalrendering-1

[27]: #translate

[28]: #translate-1

[29]: #sequenceinput

[30]: #genetableoptions

[31]: #hasnoresults

[32]: #noresults

[33]: #lockgene

[34]: #genemodel

[35]: #unlockgene

[36]: #blastdatatracker

[37]: #getmultiplegenesequences

[38]: #blastresult

[39]: #genetablerow

[40]: #getgenesequences

[41]: #blasttypes

[42]: #getreferencesequences

[43]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String

[44]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object

[45]: https://www.robinwieruch.de/gentle-introduction-higher-order-components/

[46]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function

[47]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array

[48]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean

[49]: https://stackoverflow.com/a/46854785/6573438

[50]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number

[51]: #interval

[52]: #sequenceinput

[53]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/undefined

[54]: #genemodel
